### ISSUE TITLE
gradient: silently wrong result when a cons-routed loss is differentiated at a variable-bound point

### ISSUE THREAD

**Issue by jotron-ops** [tsotchke/eshkol#340]
### Summary

`gradient` returns a **silently wrong** value when two conditions hold together:

1. the loss routes tracked values through a `cons` cell (`car`/`cdr`), **and**
2. the differentiation point is a **variable bound to a vector** (not a `#(...)` literal).

Either condition alone is fine. Only the conjunction fails — and it fails *quietly*: a wrong gradient is still returned alongside a stderr message, so a training loop keeps running on corrupt values.

### Repro

```scheme
;; L(p) = p0 + p0^2  =>  dL/dp0 = 1 + 2*p0 = 7 at p0 = 3
(define (loss-plain p) (let ((x (vector-ref p 0))) (+ x (* x x))))
(define (loss-cons p)
  (let* ((x (vector-ref p 0)) (pr (cons x (* x x)))) (+ (car pr) (cdr pr))))

(define pv (vector 3.0))                      ; VAR-bound point

(display (gradient loss-plain #(3.0)))        ; A => #(7)   OK
(display (gradient loss-cons  #(3.0)))        ; B => #(7)   OK
(display (gradient loss-plain pv))            ; C => #(7)   OK
(display (gradient loss-cons  pv))            ; D => #(1)   WRONG
```

|            | `#(...)` literal | VAR-bound |
|------------|------------------|-----------|
| plain      | `#(7)` ✓         | `#(7)` ✓  |
| via `cons` | `#(7)` ✓         | **`#(1)` ✗** |

**Actual (case D):**

```
ERROR: Attempted to get int64 from non-int-storage cell (type=6)
#(1)
```

**Expected:** `#(7)`.

The forward *value* is correct in all four cases (`12`); only the gradient is wrong.

### Why this one is nasty

It does not crash and does not raise — it returns a plausible-looking gradient. In an iterated optimizer the run simply converges to the wrong place. We only caught it because a training loop produced a stderr flood while still "working".

### Possibly related

This may share a root cause with the `hessian` VAR-point crash filed separately, since both are triggered specifically by a *variable-bound* point taking a different path than a literal one. Filing separately because the symptoms differ (silent wrong value vs. SIGSEGV).

### Workaround

Keep accumulators in named-`let` bindings rather than pairs/lists inside a differentiated loss, or pass the point as a literal.

### Environment

- eshkol `cddadded` (v1.3.3-evolve line)
- LLVM 21.1.8, macOS 26.4.1 (arm64, M-series), Release build via CMake/Ninja
